### DISCUSSION TITLE
[Feature Request] Config File

### DISCUSSION THREAD

**Discussion by MGatner** [composer-unused/composer-unused#272] [Ideas]
Hi! I would like to request the addition of config file processing. This would pare down on command line arguments needing to be passed, and would help keep details specific to each library with that repo.

Initial request for configurable items:
* Additional files to autoload prior to scanning (see: https://github.com/composer-unused/composer-unused/issues/242)
* A list of packages to exclude, same as `--excludePackage`
* Verbosity settings (`-v`, `-vv`, `-vvv`)
* Output options

***

I will put forth a $50 bounty on this request, and would be glad for anyone else to chip in!

**Comment by icanhazstring** [comment]
**Additional files to autoload prior to scanning**

This might need to come with a certain structure.
As I need to map additional files to specific packages.

E.g. You might need to configure it like this:

```
"addtionalFiles": {
    "tatter/alerts": [
        "/path/to/file/A.php"
    ]
}
```

This way I might be able to map additional files to vendor package `composer.json` so that composer-unused "knowns" that these files are from that package.

---

Or would you prefer a more "dynamic" approach?

**Reply by MGatner** [reply]
Manual mapping would be fine! But for my case these will always be Composer vendor files, so theoretically the source could be determined from the path. At the end of the day I care more about the function than the approach - I trust that to the developer. 🤗

**Reply by icanhazstring** [reply]
0.8.1 was released supporting a dedicated configuration file with the ability to add additional files to be covered per dependency :)

**Reply by MGatner** [reply]
Am I doing something wrong that Phive is still pulling an old version?
```bash
sudo phive install composer-unused
Phive 0.15.0 - Copyright (C) 2015-2022 by Arne Blankerts, Sebastian Heuer and Contributors
Linking /root/.phive/phars/composer-unused-0.7.7.phar to /home/ubuntu/tools/composer-unused
```

**Reply by MGatner** [reply]
(Good news is that `setup-php` appears to have the latest version: https://github.com/tattersoftware/codeigniter4-files/runs/5335229723?check_suite_focus=true#step:3:23)

**Comment by icanhazstring** [comment]
**Output options**

What kind of output options are you thinking of?

I can think of
- json
- xml

Any certain output format? (JUnit)

**Reply by MGatner** [reply]
This is not really a priority for me, I was more brainstorming things that could go in a Config file. One output format that would be very slick to have is GitHub Annotations. I'm not sure how that would work for general errors (since you aren't targeting any code or files specifically) - maybe it would go on composer.json?

**Reply by icanhazstring** [reply]
Possible output:

https://github.com/FriendsOfPHP/PHP-CS-Fixer/blob/v3.0.2/doc/schemas/fix/checkstyle.xsd

**Reply by icanhazstring** [reply]
Could be reported using
https://github.com/staabm/annotate-pull-request-from-checkstyle

**Reply by icanhazstring** [reply]
And 0.8.1 release with github annotation support. More report styles could be added. Just file a new feature request :)

**Reply by MGatner** [reply]
YAY! Exciting! I am digging back out from traveling but I will dig into this over the weekend.

**Reply by MGatner** [reply]
Using the latest **bin/composer-unused** from `main` I'm not getting the GitHub format locally. Does `CiDetector` override CLI arguments?
```bash
/opt/codeigniter4/files$ ~/composer-unused/bin/composer-unused --output-format=github


Results
-------

Found 6 used, 2 unused, 0 ignored and 0 zombie packages

 Used packages
 ✓ php
 ✓ codeigniter4/authentication-implementation (required by: tatter/permits)
 ✓ tatter/exports
 ✓ tatter/frontend
 ✓ tatter/permits
 ✓ tatter/thumbnails

 Unused packages
 ✗ enyo/dropzone
 ✗ tatter/preferences

 Ignored packages

 Zombies exclusions (did not match any package)
```

**Reply by icanhazstring** [reply]
It shouldn't but seems like. Can you create a new issue linking to this? :)

**Comment by MGatner** [comment]
Very awesome solution! I was able to define a config file easily based off your example:
```php
<?php

declare(strict_types=1);

use ComposerUnused\ComposerUnused\Configuration\Configuration;
use ComposerUnused\ComposerUnused\Configuration\NamedFilter;

return static function (Configuration $config): Configuration {
    return $config
        ->addNamedFilter(NamedFilter::fromString('enyo/dropzone'))
        ->setAdditionalFilesFor('tatter/preferences', [
            __DIR__ . '/vendor/tatter/preferences/src/Helpers/preferences_helper.php',
        ]);
};
```

This trimmed down my CLI command down to just the executable: `~/composer-unused/bin/composer-unused`

And the end result was exactly as expected, beautifully colored and formatted:
<img width="876" alt="Screen Shot 2022-02-25 at 11 16 25 AM" src="https://user-images.githubusercontent.com/17572847/155749787-9012bbab-a10b-4693-bd48-24526cf1724a.png">

Bounty on its way :)

**Reply by icanhazstring** [reply]
Glad it worked out. And thanks again for your donations 🤩

**Reply by MGatner** [reply]
Absolutely! Would you prefer a PayPal one-time or GitHub $5/month x 10 months?

**Reply by icanhazstring** [reply]
Depending on how you would like to pay. I don't mind either one 😉

**Reply by MGatner** [reply]
I like showing as a Sponsor. There's always the chance I will forget to cancel it down the road as well 🤭
